climada/hazard/base.py

```diff
@@ -369,6 +371,361 @@ def set_vector(self, *args, **kwargs):
                        "Use Hazard.from_vector instead.")
         self.__dict__ = Hazard.from_vector(*args, **kwargs).__dict__
 
+    @classmethod
+    def from_raster_xarray(
+        cls,
+        data: Union[xr.Dataset, str],
+        *,
+        intensity: str = "intensity",
+        coordinate_vars: Optional[Dict[str, str]] = None,
+        data_vars: Optional[Dict[str, str]] = None,
+        hazard_type: str = "",
+        intensity_unit: str = "",
+    ):
+        """Read raster-like data from an xarray Dataset or a raster data file
+
+        This method reads data that can be interpreted using three coordinates for time,
+        latitude, and longitude. The data and the coordinates themselves may be organized
+        in arbitrary dimensions in the Dataset (e.g. three dimensions 'year', 'month',
+        'day' for the coordinate 'date' representing time). The three coordinates to be
+        read can be specified via the ``coordinate_vars`` parameter.
+
+        The only required data is the intensity. For all other data, this method can
+        supply sensible default values. By default, this method will try to find these
+        "optional" data in the Dataset and read it, or use the default values otherwise.
+        Users may specify the variables in the Dataset to be read for certain Hazard
+        object entries, or may indicate that the default values should be used although
+        the Dataset contains appropriate data. This behavior is controlled via the
+        ``data_vars`` parameter.
+
+        If this method succeeds, it will always return a "consistent" Hazard object,
+        meaning that the object can be used in all CLIMADA operations without throwing
+        an error due to missing data or faulty data types.
+
+        Notes
+        -----
+        * Intensity and fraction data must be three-dimensional data that is interpreted
+          with the coordinates given by ``coordinate_vars`` (or the default). All other
+          data must be given in time coordinates only.
+        * To avoid confusion in the call signature, all parameters are keyword-only
+          arguments, except ``data``.
+        * The attributes ``Hazard.tag.haz_type`` and ``Hazard.unit`` currently cannot be
+          read from the Dataset. Use the method parameters to set these attributes or set
+          them in the resulting ``Hazard`` object by yourself.
+        * This method **does not** read lazily. Single data arrays must fit into memory.
```

Comment:
That is probably a big issue. The intensity and fraction are sparse, thus casting them to dense will most likely make this method not usable for most larger datasets (which are quite common in CLIMADA). Any idea how to solve this?

Reply:
I would say it depends. I stopped at some point of my own wishlist because it otherwise get's too much to take into account in a single PR. I personally would say that this can be solved in a follow-up PR. There is quite a lot one has to take into account here, in particular the chunking of dask arrays and a memory-efficient conversion of those into a CSR matrix. I haven't tried that out yet.

Reply:
Ok, fair enough. This will probably very strongly limit the use of this method for CLIMADA, so we should definitely keep it on the radar.